pkg/controller/remotemachineset/vsphereactuator.go

```diff
@@ -0,0 +1,74 @@
+package remotemachineset
+
+import (
+	"fmt"
+
+	machineapi "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	installvsphere "github.com/openshift/installer/pkg/asset/machines/vsphere"
+	installertypes "github.com/openshift/installer/pkg/types"
+	installertypesvsphere "github.com/openshift/installer/pkg/types/vsphere"
+)
+
+// VSphereActuator encapsulates the pieces necessary to be able to generate
+// a list of MachineSets to sync to the remote cluster
+type VSphereActuator struct {
+	logger log.FieldLogger
+}
+
+var _ Actuator = &VSphereActuator{}
+
+// NewVSphereActuator is the constructor for building a VSphereActuator
+func NewVSphereActuator(logger log.FieldLogger) (*VSphereActuator, error) {
+	actuator := &VSphereActuator{
+		logger: logger,
+	}
+	return actuator, nil
+}
+
+// GenerateMachineSets satisfies the Actuator interface and will take a clusterDeployment and return a list of MachineSets
+// to sync to the remote cluster.
+func (a *VSphereActuator) GenerateMachineSets(cd *hivev1.ClusterDeployment, pool *hivev1.MachinePool, logger log.FieldLogger) ([]*machineapi.MachineSet, bool, error) {
+	if cd.Spec.ClusterMetadata == nil {
+		return nil, false, errors.New("ClusterDeployment does not have cluster metadata")
+	}
+	if cd.Spec.Platform.VSphere == nil {
+		return nil, false, errors.New("ClusterDeployment is not for VSphere")
+	}
+	if pool.Spec.Platform.VSphere == nil {
+		return nil, false, errors.New("MachinePool is not for VSphere")
+	}
+
+	computePool := baseMachinePool(pool)
+	computePool.Platform.VSphere = &installertypesvsphere.MachinePool{
+		NumCPUs:           pool.Spec.Platform.VSphere.NumCPUs,
+		NumCoresPerSocket: pool.Spec.Platform.VSphere.NumCoresPerSocket,
+		MemoryMiB:         pool.Spec.Platform.VSphere.MemoryMiB,
+		OSDisk: installertypesvsphere.OSDisk{
+			DiskSizeGB: pool.Spec.Platform.VSphere.DiskSizeGB,
+		},
+	}
+
+	ic := &installertypes.InstallConfig{
+		Platform: installertypes.Platform{
+			VSphere: &installertypesvsphere.Platform{},
```

Comment:
This is missing a lot of fields that are used by the installer to create the MachineSet. For example, it is missing the vCenter. Are those fields somehow obtained in-cluster and not needed?

https://github.com/openshift/installer/blob/8da3ffa23277b55b3ddbf53761063d22a30ce8a2/pkg/asset/machines/vsphere/machines.go#L66.

Reply:
For this, I copied the pattern used by the [OpenStack actuator](https://github.com/openshift/hive/blob/master/pkg/controller/remotemachineset/openstackactuator.go#L66-L75), which is basically only fill out the bare minimum used to be able to generate the machinesets.

When I originally looked through the `remotemachineset` controller code, I didn't see any place these values were used. But, that was because I was focusing on the remotemachineset Update code.

During a remotemachineset create call, this information is used, so I think you're correct that we should be filling this out.

I'm running some tests now to confirm.

@dgoodwin do you think that this is a problem with remotemachineset create in the OpenStack actuator as well?

Reply:
For posterity, the decision here was to fill out the install config with enough information so that the created machinesets had the same critical data that the machinesets created during install had. This was manually verified by comparing the data structures.